internal_task_handlers.go

```diff
@@ -472,17 +472,58 @@ ProcessEvents:
 	return taskCompletionRequest, stackTrace, nil
 }
 
+func isVersionMarkerDecision(d *s.Decision) bool {
+	if d.GetDecisionType() == s.DecisionType_RecordMarker &&
+		d.RecordMarkerDecisionAttributes.GetMarkerName() == versionMarkerName {
+		return true
+	}
+	return false
+}
+
+func isVersionMarkerEvent(e *s.HistoryEvent) bool {
+	if e.GetEventType() == s.EventType_MarkerRecorded &&
+		e.MarkerRecordedEventAttributes.GetMarkerName() == versionMarkerName {
+		return true
+	}
+	return false
+}
+
 func matchReplayWithHistory(replayDecisions []*s.Decision, historyEvents []*s.HistoryEvent) error {
-	for i := 0; i < len(historyEvents); i++ {
-		e := historyEvents[i]
-		if i >= len(replayDecisions) {
+	di := 0
+	hi := 0
+	hSize := len(historyEvents)
+	dSize := len(replayDecisions)
+matchLoop:
+	for hi < hSize || di < dSize {
+		var e *s.HistoryEvent
+		if hi < hSize {
+			e = historyEvents[hi]
+		}
+		if isVersionMarkerEvent(e) {
```

Comment:
not fully follow the intention of the change here, but looks like you are trying to skip comparing versionMarkerEvent and versionMarkerDecision. If that is the case, I think it is clearer to avoid adding these 2 types into the replayDecisions and historyEvents at first place, then you don't need any change to matchReplayWithHistory.